File: backend/lenspirecrm/env_validator.py
import os
import sys
from pathlib import Path
from typing import List, Tuple

BASE_DIR = Path(__file__).resolve().parent.parent

REQUIRED_VARS = [
    ("SECRET_KEY", "Django secret key for session signing", True),
    ("ALLOWED_HOSTS", "Comma-separated list of allowed hosts", False),
    ("CORS_ALLOWED_ORIGINS", "Comma-separated list of allowed CORS origins", False),
    ("BACKUP_ENCRYPTION_KEY", "32-character key for backup encryption", True),
]

CONDITIONAL_VARS = [
    ("POSTGRES_DB", "PostgreSQL database name", "DATABASE_URL not set"),
    ("POSTGRES_USER", "PostgreSQL database user", "DATABASE_URL not set"),
    ("POSTGRES_PASSWORD", "PostgreSQL database password", "DATABASE_URL not set"),
    ("POSTGRES_HOST", "PostgreSQL host", "DATABASE_URL not set"),
    ("POSTGRES_PORT", "PostgreSQL port", "DATABASE_URL not set"),
    ("CELERY_BROKER_URL", "Celery broker URL for async tasks", "CELERY tasks enabled"),
]

OPTIONAL_VARS = [
    ("DEBUG", "Debug mode (default: false)", False),
    ("TIME_ZONE", "Server timezone (default: UTC)", False),
    ("SENTRY_DSN", "Sentry error tracking DSN", False),
    ("GOOGLE_CLIENT_ID", "Google OAuth client ID for Drive integration", False),
    ("GOOGLE_CLIENT_SECRET", "Google OAuth client secret", False),
    ("DRIVE_TOKEN_KEY", "32-character key for Drive token encryption", False),
]


def validate_env():
    """
    Validate required environment variables on startup.
    
    Returns:
        Tuple of (is_valid: bool, missing_vars: List[str], warnings: List[str])
    """
    errors = []
    warnings = []
    
    debug = os.getenv("DEBUG", "true").lower() == "true"
    
    for var_name, description, required_in_prod in REQUIRED_VARS:
        value = os.getenv(var_name, "")
        
        if not value:
            if required_in_prod and not debug:
                errors.append(f"  [X] {var_name}: {description} (REQUIRED in production)")
            elif var_name == "SECRET_KEY":
                if value == "lenspire-development-key-change-before-production":
                    warnings.append(f"  [!] {var_name}: Using default development key")
            elif required_in_prod:
                warnings.append(f"  [!] {var_name}: {description} (missing)")
        else:
            if var_name == "BACKUP_ENCRYPTION_KEY" and len(value) != 32:
                errors.append(f"  [X] {var_name}: Must be exactly 32 characters (got {len(value)})")
            elif var_name == "SECRET_KEY" and len(value) < 32:
                warnings.append(f"  [!] {var_name}: Should be at least 32 characters for security")
    
    database_url = os.getenv("DATABASE_URL", "")
    if not database_url:
        for var_name, description, condition in CONDITIONAL_VARS:
            if condition == "DATABASE_URL not set":
                value = os.getenv(var_name, "")
                if not value and not debug:
                    errors.append(f"  [X] {var_name}: {description} (REQUIRED when {condition})")
    
    celery_disabled = os.getenv("CELERY_TASK_ALWAYS_EAGER", "false").lower() == "true"
    if not celery_disabled and not os.getenv("CELERY_BROKER_URL"):
        warnings.append("  [!] CELERY_BROKER_URL: Celery tasks will run synchronously without broker")
    
    return len(errors) == 0, errors, warnings


def check_settings():
    """
    Validate Django settings and print a startup report.
    
    Call this from settings.py after all settings are defined.
    """
    debug = os.getenv("DEBUG", "true").lower() == "true"
    
    print("\n" + "=" * 70)
    print("  LenspireCRM Backend - Environment Validation")
    print("=" * 70)
    
    if debug:
        print("  [!] RUNNING IN DEBUG MODE - Not for production use")
    
    print()
    
    is_valid, errors, warnings = validate_env()
    
    if errors:
        print("  ERRORS (must fix before deployment):")
        for error in errors:
            print(error)
        print()
    
    if warnings:
        print("  WARNINGS:")
        for warning in warnings:
            print(warning)
        print()
    
    if is_valid and not errors:
        print("  [OK] All required environment variables are set")
        if warnings:
            print("  [!] Some optional variables have warnings (see above)")
        print()
    
    print("=" * 70)
    
    if not is_valid and not debug:
        print("\n  [STOP] Fatal: Missing required environment variables")
        print("  Set the missing variables and restart the server.\n")
        sys.exit(1)
    elif errors and not debug:
        sys.exit(1)


if __name__ == "__main__":
    from dotenv import load_dotenv
    load_dotenv(BASE_DIR / ".env")
    check_settings()

File: src/main/env-validator.js
const path = require('path');
const fs = require('fs');

const REQUIRED_VARS = [
    { name: 'LENSPIRE_DB_PATH', description: 'Path to SQLite database file', required: false, default: './src/database/tracker.db' },
    { name: 'LENSPIRE_CLOUD_API_URL', description: 'Cloud API endpoint URL', required: false, default: 'https://crm.lenspireai.com' },
    { name: 'LENSPIRE_BACKUP_PASSWORD', description: 'Password for encrypted backups', required: false, sensitive: true },
];

const PROD_ONLY_VARS = [
    { name: 'CSC_LINK', description: 'Code signing certificate (base64)', required: false, sensitive: true },
    { name: 'CSC_KEY_PASSWORD', description: 'Code signing certificate password', required: false, sensitive: true },
];

function validateEnv() {
    const errors = [];
    const warnings = [];
    const isDev = process.env.NODE_ENV !== 'production';
    
    console.log('\n' + '='.repeat(70));
    console.log('  LenspireCRM Pro (Electron) - Environment Validation');
    console.log('='.repeat(70));
    console.log();
    
    if (isDev) {
        console.log('  ℹ Running in DEVELOPMENT mode\n');
    } else {
        console.log('  ⚠ RUNNING IN PRODUCTION MODE\n');
    }
    
    for (const varDef of REQUIRED_VARS) {
        const value = process.env[varDef.name];
        
        if (!value) {
            if (varDef.required) {
                errors.push(`  ✗ ${varDef.name}: ${varDef.description} (REQUIRED)`);
            } else if (varDef.default) {
                console.log(`  ℹ ${varDef.name}: Using default: ${varDef.default}`);
            } else {
                warnings.push(`  ⚠ ${varDef.name}: ${varDef.description} (optional, not set)`);
            }
        } else {
            if (varDef.sensitive) {
                console.log(`  ✓ ${varDef.name}: [REDACTED]`);
            } else {
                const displayValue = value.length > 50 ? value.substring(0, 50) + '...' : value;
                console.log(`  ✓ ${varDef.name}: ${displayValue}`);
            }
            
            if (varDef.name === 'LENSPIRE_DB_PATH') {
                const dbPath = path.resolve(value);
                if (!fs.existsSync(dbPath)) {
                    warnings.push(`  ⚠ ${varDef.name}: Database file does not exist at ${dbPath}`);
                }
            }
        }
    }
    
    if (!isDev) {
        console.log('\n  Production-specific checks:\n');
        for (const varDef of PROD_ONLY_VARS) {
            const value = process.env[varDef.name];
            if (!value && varDef.required) {
                errors.push(`  ✗ ${varDef.name}: ${varDef.description} (REQUIRED for production builds)`);
            } else if (!value) {
                warnings.push(`  ⚠ ${varDef.name}: ${varDef.description} (recommended for signed builds)`);
            } else {
                console.log(`  ✓ ${varDef.name}: [REDACTED]`);
            }
        }
    }
    
    console.log();
    
    if (errors.length > 0) {
        console.log('  ERRORS (must fix):');
        errors.forEach(err => console.log(err));
        console.log();
    }
    
    if (warnings.length > 0) {
        console.log('  WARNINGS:');
        warnings.forEach(warn => console.log(warn));
        console.log();
    }
    
    if (errors.length === 0) {
        console.log('  ✓ Environment validation passed');
        if (warnings.length > 0) {
            console.log('  ⚠ Some warnings present (see above)');
        }
    }
    
    console.log('='.repeat(70) + '\n');
    
    if (errors.length > 0 && !isDev) {
        console.error('  🛑 Fatal: Missing required environment variables\n');
        process.exit(1);
    }
    
    return errors.length === 0;
}

function validateBeforeStart() {
    try {
        const isValid = validateEnv();
        if (!isValid && process.env.NODE_ENV === 'production') {
            throw new Error('Environment validation failed');
        }
        return isValid;
    } catch (error) {
        console.error('  🛑 Environment validation error:', error.message);
        if (process.env.NODE_ENV === 'production') {
            process.exit(1);
        }
        return false;
    }
}

module.exports = { validateEnv, validateBeforeStart };

if (require.main === module) {
    require('dotenv').config({ path: path.join(__dirname, '..', '.env') });
    validateEnv();
}
